Draft of episode 1: Intro to Raster Data with Python

rbavery opened the pull request:
This lesson mostly follows the structure of the [corresponding R lesson](https://datacarpentry.org/r-raster-vector-geospatial/01-raster-structure/index.html) with a few notable changes:

- The No Data section has been adapted to go into more detail on profiling no data values and how to mask these values for both plotting and calculation with `numpy.ma.masked_where`. Code examples are included that generate the plots.

- I think the Bad Data Values section from the R lesson would be better placed in the next episode on plotting, since this episode is already long and covers each of the metadata attributes included in the PROJ string.

Reviews appreciated! I'd particularly like a look at the coordinate reference system example, I couldn't find a way to generate the PROJ4 string using only rasterio and it includes the most information in a single line. The alternative is accessing the projection, unit, zone, datum information through various attributes on the `meta` attribute so I thought this was cleaner. the `meta` attribute's `to_proj4()` function does not return a full PROJ string. I'd also really like feedback on the No Data section since that is what I reworked the most. Any other comments are welcome as well.

For those new to creating and editing lesson content (I know I am), you can view the result of this pull request at my fork of this repo at: https://rbavery.github.io/geospatial-python/01-raster-structure/index.html

rbavery commented:
Hi @dvanic here's a draft of episode 1 if you have time to review.

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+~~~
+{: .output}
+
+You can convert the EPSG code to a PROJ4 string with `earthpy.epsg`, another python `dict` which maps epsg codes (keys) to
```
This is somewhat convoluted/confusing, as you need to manually type the code from the crs created by rasterio to the string expected by earthpy.epsg. 

Should we instead be building up to:
```
earthpy.epsg[str(surface_model_HARV.crs.to_epsg())]
```
?

Perhaps by showing that you need a string input for earthpy.epsg, and then showing how to get the string from rasterio?

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+>
+> > ## Answers
+> > `+units=m` tells us that our data is in meters.
+> > We could also get this information from the attribute `surface_model.crs.linear_units`.
```
`surface_model_HARV`

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+You could also get each of these values one by one using `numpy`. What if we wanted to calculate 25% and 75% quartiles?
+
+~~~
+import numpy
```
Should we `import numpy as np` here? This seems to be standard(ish)?

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+
+> ## Data Tip - Set min and max values
+> You may notice that `numpy.percentile` didn't require an `axis=None` argument. This is because `axis=None` is the default for most numpy 
+> functions. It's always good to check out the docs on a function to see what the default argumetns are, particularly when working with 
```
argumetns -> arguments

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+there is only 1 band in a raster, it will be placed in it's own band axis.
+
+~~~
+surface_model_HARV_arr_3D = surface_model.read()
```
`surface_model` -> `surface_model_HARV`

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+Earlier, we read our raster in as a 2D array instead
+
+~~~
+surface_model_HARV_arr_2D = surface_model.read(1)
```
`surface_model` -> `surface_model_HARV`

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+
+## Dealing with Missing Data
+
+Raster data often has a no data value associated with it. This is a value
```
no data value -> 'no data value'

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+
+<img src="../fig/01-demonstrate-no-data-black-3.png" title="plot of chunk demonstrate-no-data-black" alt="plot of chunk demonstrate-no-data-black" width="612" style="display: block; margin: auto;" />
+
+`rasterio` assigns a specific number as missing data to the `meta` attribute when the dataset is read, base don the file's own metadata. While the GeoTiff's `nodata` attribute is assigned to the value `-1.7e+308`, it turns out the missing data at the edges are represented by the value `0`. In order to run calculations on this image that ignore these edge values or plot he iamge without the nodata ...
```
base don -> based on

he iamge -> the image

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+`rasterio` assigns a specific number as missing data to the `meta` attribute when the dataset is read, base don the file's own metadata. While the GeoTiff's `nodata` attribute is assigned to the value `-1.7e+308`, it turns out the missing data at the edges are represented by the value `0`. In order to run calculations on this image that ignore these edge values or plot he iamge without the nodata ...
+
+In the next image, the black edges have been masked using `numpy.ma.masked_where()`, a function that assigns no data values where a condition is true.
+
```
You use numpy.ma here (and imported numpy), but then below use np.ma.masked_where()... We should probably check for consistency later.

rbavery replied:
Made this consistent, thanks for the catch.

dvanic commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+If your raster already has `nodata` values set correctly but you aren't sure where they are, you can deliberately plot them in a particular colour. This can be useful when checking a dataset's coverage. For instance, sometimes data can be missing where a sensor could not 'see' its target data, and you may wish to locate that missing data and fill it in. With Python, we can plot a boolean array of ...
+
+~~~
+earthpy.plot.plot_rgb(
```
This code cries about two issues:

1. `name 'rgb_HARV_nan_arr' is not defined`
2. If you replace that with `rgb_HARV_masked_arr`, then `name 'ax' is not defined`

rbavery replied:
Fixed, thanks for catching this. I tested the whole lesson again from scratch and it works.

rbavery commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+This object has an attribute, `.meta`, which contains the metadata for the file we opened.
+
+
+This metadata is stored in the form of a new data type, a `dict`, or dictionary. Dictionaries store
```
Yeah I agree. Like the R lesson we should have a separate repo for an abridged intro to python lesson that can be taught before this geospatial python lesson. @ErinBecker can we setup another repo named "python-intro-geospatial". We could condense the gapminder lesson similar to how this R lesson condenses R for Reproducible Data Science: https://github.com/datacarpentry/r-intro-geospatial

rbavery commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+The first call to `rasterio.open()` opens the file and returns an object that we store in a variable, `surface_model_HARV`.
+This object has an attribute, `.meta`, which contains the metadata for the file we opened.
+
+
```
Thanks, good suggestion.

The transform encodes how pixel coordinates are mapped to spatial coordinates, so it contains a matrix representation of a spatial projection that allows for projections that don't have EPSG codes. R reports the PROJ4 string if I remember correctly, this PROJ4 string is also accessible from the CRS object and we examine the PROJ4 string representation later in the lesson. I think rasterio chose to use the EPSG code as the representation of projections in the .meta dictionary instead of a proj4 string because some people prefer to work directly with EPSG codes.

rbavery replied:
I included a short description of each metadata value, including the transform

rbavery commented on _episodes/01-raster-structure.md:
```diff
@@ -15,10 +15,474 @@ objectives:
+`rasterio` for reading and writing rasters, `scipy` for descriptive statistics, and `earthpy.plot` for plotting. Make sure that you have these packages installed and imported.
+
+
+~~~
```
Done! all libraries imported at the start.

rbavery commented:
@dvanic Thanks a bunch Darya for the thorough review. I'm getting used to contributing to lessons and accidentally pushed this recent commit to address your review to this repo instead of my fork, which triggered a merge. I think I addressed all your comments but if you see something I can open another PR.

I think it's a good idea to break up the work into separate branches, in the future I'll make a separate branch on this repo to track the work rather than using gh-pages, that way folks won't step on each other's toes when merging to this repo's gh-pages.
